# Data Analysis on Hacker news posts

In this project, we'll compare two different types of posts from [Hacker News](https://news.ycombinator.com/). The two types of posts we'll explore begin with either `Ask HN` or `Show HN`.

Users submit `Ask HN` posts to ask the Hacker News community a specific question, such as "What is the best online course you've ever taken?" Likewise, users submit `Show HN` posts to show the Hacker News community a project, product, or just generally something interesting.

We'll specifically compare these two types of posts to determine the following:

- Do `Ask HN` or `Show HN` receive more comments on average?
- Do posts created at a certain time receive more comments on average?

It should be noted that the data set we're working with was reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions.

# Introduction

First, we'll read in the data and remove the headers.

In [34]:
# Read in the data.
import csv

f = open('hacker_news.csv')
hn = list(csv.reader(f))
hn[:5]

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'],
 ['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01']]

In [35]:
# Removing Headers from a List of Lists.
headers = hn[0]
hn = hn[1:]
print(headers)
print(hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


We can see above that the data set contains the title of the posts, the number of comments for each post, and the date the post was created. Let's start by exploring the number of comments for each type of post. 

# Extracting Ask HN and Show HN Posts

First, we'll identify posts that begin with either `Ask HN` or `Show HN` and separate the data for those two types of posts into different lists. Separating the data makes it easier to analyze in the following steps.

In [36]:
# Identify posts that begin with either `Ask HN` or `Show HN` and separate the data into different lists.
ask_posts = []
show_posts =[]
other_posts = []

for post in hn:
    title = post[1]
    if title.lower().startswith("ask hn"):
        ask_posts.append(post)
    elif title.lower().startswith("show hn"):
        show_posts.append(post)
    else:
        other_posts.append(post)
        
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1744
1162
17194


# Calculating the Average Number of Comments for Ask HN and Show HN Posts

Now that we separated ask posts and show posts into different lists, we'll calculate the average number of comments each type of post receives.

In [37]:
# Calculate the average number of comments `Ask HN` posts receive.
total_ask_comments = 0

for post in ask_posts:
    total_ask_comments += int(post[4])
    
avg_ask_comments = total_ask_comments / len(ask_posts)
print(avg_ask_comments)

14.038417431192661


In [38]:
# Calculate the average number of comments `Show HN` posts receive.
total_show_comments = 0

for post in show_posts:
    total_show_comments += int(post[4])
    
avg_show_comments = total_show_comments / len(show_posts)
print(avg_show_comments)

10.31669535283993


On average, ask posts in our sample receive approximately 14 comments, whereas show posts receive approximately 10. Since ask posts are more likely to receive comments, we'll focus our remaining analysis just on these posts. 

# Finding the Amount of Ask Posts and Comments by Hour Created

Next, we'll determine if we can maximize the amount of comments an ask post receives by creating it at a certain time. First, we'll find the amount of ask posts created during each hour of day, along with the number of comments those posts received. Then, we'll calculate the average amount of comments ask posts created at each hour of the day receive.

In [39]:
# Calculate the amount of ask posts created during each hour of day and the number of comments received.
import datetime as dt

result_list = []

for post in ask_posts:
    result_list.append(
        [post[6], int(post[4])]
    )

comments_by_hour = {}
counts_by_hour = {}
date_format = "%m/%d/%Y %H:%M"

for each_row in result_list:
    date = each_row[0]
    comment = each_row[1]
    time = dt.datetime.strptime(date, date_format).strftime("%H")
    if time in counts_by_hour:
        comments_by_hour[time] += comment
        counts_by_hour[time] += 1
    else:
        comments_by_hour[time] = comment
        counts_by_hour[time] = 1

comments_by_hour

{'09': 251,
 '13': 1253,
 '10': 793,
 '14': 1416,
 '16': 1814,
 '23': 543,
 '12': 687,
 '17': 1146,
 '15': 4477,
 '21': 1745,
 '20': 1722,
 '02': 1381,
 '18': 1439,
 '03': 421,
 '05': 464,
 '19': 1188,
 '01': 683,
 '22': 479,
 '08': 492,
 '04': 337,
 '00': 447,
 '06': 397,
 '07': 267,
 '11': 641}

# Calculating the Average Number of Comments for Ask HN Posts by Hour

In [40]:
# Calculate the average amount of comments `Ask HN` posts created at each hour of the day receive.
avg_by_hour = []

for hr in comments_by_hour:
    avg_by_hour.append([hr, comments_by_hour[hr] / counts_by_hour[hr]])

avg_by_hour

[['09', 5.5777777777777775],
 ['13', 14.741176470588234],
 ['10', 13.440677966101696],
 ['14', 13.233644859813085],
 ['16', 16.796296296296298],
 ['23', 7.985294117647059],
 ['12', 9.41095890410959],
 ['17', 11.46],
 ['15', 38.5948275862069],
 ['21', 16.009174311926607],
 ['20', 21.525],
 ['02', 23.810344827586206],
 ['18', 13.20183486238532],
 ['03', 7.796296296296297],
 ['05', 10.08695652173913],
 ['19', 10.8],
 ['01', 11.383333333333333],
 ['22', 6.746478873239437],
 ['08', 10.25],
 ['04', 7.170212765957447],
 ['00', 8.127272727272727],
 ['06', 9.022727272727273],
 ['07', 7.852941176470588],
 ['11', 11.051724137931034]]

# Sorting and Printing Values from a List of Lists

In [41]:
swap_avg_by_hour = []

for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])
    
print(swap_avg_by_hour)

sorted_swap = sorted(swap_avg_by_hour, reverse=True)

sorted_swap

[[5.5777777777777775, '09'], [14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [16.796296296296298, '16'], [7.985294117647059, '23'], [9.41095890410959, '12'], [11.46, '17'], [38.5948275862069, '15'], [16.009174311926607, '21'], [21.525, '20'], [23.810344827586206, '02'], [13.20183486238532, '18'], [7.796296296296297, '03'], [10.08695652173913, '05'], [10.8, '19'], [11.383333333333333, '01'], [6.746478873239437, '22'], [10.25, '08'], [7.170212765957447, '04'], [8.127272727272727, '00'], [9.022727272727273, '06'], [7.852941176470588, '07'], [11.051724137931034, '11']]


[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [13.20183486238532, '18'],
 [11.46, '17'],
 [11.383333333333333, '01'],
 [11.051724137931034, '11'],
 [10.8, '19'],
 [10.25, '08'],
 [10.08695652173913, '05'],
 [9.41095890410959, '12'],
 [9.022727272727273, '06'],
 [8.127272727272727, '00'],
 [7.985294117647059, '23'],
 [7.852941176470588, '07'],
 [7.796296296296297, '03'],
 [7.170212765957447, '04'],
 [6.746478873239437, '22'],
 [5.5777777777777775, '09']]

In [42]:
# Sort the values and print the the 5 hours with the highest average comments.

print("Top 5 Hours for 'Ask HN' Comments")
for avg, hr in sorted_swap[:5]:
    print(
        "{}: {:.2f} average comments per post".format(
            dt.datetime.strptime(hr, "%H").strftime("%H:%M"),avg
        )
    )

Top 5 Hours for 'Ask HN' Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest 
average number of comments.

According to the data set [documentation](https://www.kaggle.com/hacker-news/hacker-news-posts/home), the timezone used is Eastern Time in the US. So, we could also write 15:00 as 3:00 pm est.

# Additional steps



* Determine if show or ask posts receive more points on average.
* Determine if posts created at a certain time are more likely to receive more points.
* Compare your results to the average number of comments and points other posts receive.

# Calculating the Average Number of Points for Ask HN and Show HN Posts

In [43]:
# Calculate the average number of points `Ask HN` posts receive.
total_ask_points = 0

for post in ask_posts:
    total_ask_points += int(post[3])
    
avg_ask_points =  total_ask_points / len(ask_posts)   
print(avg_ask_points)

15.061926605504587


In [44]:
# Calculate the average number of points `show HN` posts receive.
total_show_points = 0

for post in show_posts:
    total_show_points += int(post[3])
    
avg_show_points =  total_show_points / len(show_posts)   
print(avg_show_points)

27.555077452667813


# Finding the Amount of Show Posts and Points by Hour Created

In [47]:
# Calculate the amount of ask posts created during each hour of day and the number of points received.
import datetime as dt

results_list = []

for post in show_posts:
    results_list.append([post[6], int(post[3])])

points_by_hour = {}
counts_by_hour = {}
date_format = "%m/%d/%Y %H:%M"

for each_row in results_list:
    date = each_row[0]
    point = each_row[1]
    time = dt.datetime.strptime(date, date_format).strftime("%H")
    if time in count_by_hour:
        points_by_hour[time] += point
        counts_by_hour[time] += 1
    else:
        points_by_hour[time] = point
        counts_by_hour[time] = 1

points_by_hour

{'14': 25,
 '22': 14,
 '18': 21,
 '07': 4,
 '20': 2,
 '05': 4,
 '16': 12,
 '19': 1,
 '15': 22,
 '03': 2,
 '17': 1,
 '06': 8,
 '02': 6,
 '13': 4,
 '08': 18,
 '21': 66,
 '04': 6,
 '11': 6,
 '12': 89,
 '23': 36,
 '09': 2,
 '01': 3,
 '10': 4,
 '00': 2}

# Calculating the Average Number of points for Show HN Posts by Hour

In [48]:
# Finding the Average Number of points for Show HN Posts by Hour
avg_pts_by_hr = []

for hr in points_by_hour:
    avg_pts_by_hr.append([hr, points_by_hour[hr] / counts_by_hour[hr]])
    
avg_pts_by_hr

[['14', 25.0],
 ['22', 14.0],
 ['18', 21.0],
 ['07', 4.0],
 ['20', 2.0],
 ['05', 4.0],
 ['16', 12.0],
 ['19', 1.0],
 ['15', 22.0],
 ['03', 2.0],
 ['17', 1.0],
 ['06', 8.0],
 ['02', 6.0],
 ['13', 4.0],
 ['08', 18.0],
 ['21', 66.0],
 ['04', 6.0],
 ['11', 6.0],
 ['12', 89.0],
 ['23', 36.0],
 ['09', 2.0],
 ['01', 3.0],
 ['10', 4.0],
 ['00', 2.0]]

# Sorting and Printing Values from a List of Lists

In [49]:
swap_avg_by_hour = []

for row in avg_pts_by_hr:
    swap_avg_by_hour.append([row[1], row[0]])
    
print(swap_avg_by_hour)

sorted_swap = sorted(swap_avg_by_hour, reverse=True)

sorted_swap

[[25.0, '14'], [14.0, '22'], [21.0, '18'], [4.0, '07'], [2.0, '20'], [4.0, '05'], [12.0, '16'], [1.0, '19'], [22.0, '15'], [2.0, '03'], [1.0, '17'], [8.0, '06'], [6.0, '02'], [4.0, '13'], [18.0, '08'], [66.0, '21'], [6.0, '04'], [6.0, '11'], [89.0, '12'], [36.0, '23'], [2.0, '09'], [3.0, '01'], [4.0, '10'], [2.0, '00']]


[[89.0, '12'],
 [66.0, '21'],
 [36.0, '23'],
 [25.0, '14'],
 [22.0, '15'],
 [21.0, '18'],
 [18.0, '08'],
 [14.0, '22'],
 [12.0, '16'],
 [8.0, '06'],
 [6.0, '11'],
 [6.0, '04'],
 [6.0, '02'],
 [4.0, '13'],
 [4.0, '10'],
 [4.0, '07'],
 [4.0, '05'],
 [3.0, '01'],
 [2.0, '20'],
 [2.0, '09'],
 [2.0, '03'],
 [2.0, '00'],
 [1.0, '19'],
 [1.0, '17']]

In [51]:
# Sort the values and print the 5 hours with the highest average points.

print("Top 5 Hours for 'Show HN' points")
for avg, hr in sorted_swap[:5]:
    print(
        "{}: {:.2f} average points per post".format(
            dt.datetime.strptime(hr, "%H").strftime("%H:%M"),avg
        )
    )

Top 5 Hours for 'Show HN' points
12:00: 89.00 average points per post
21:00: 66.00 average points per post
23:00: 36.00 average points per post
14:00: 25.00 average points per post
15:00: 22.00 average points per post


# Conclusion

In this project, we analyzed ask posts and show posts to determine which type of post and time receive the most comments and points on average. Based on this analysis, we concluded that posting ask post on hacker news at 15:00 (3:00 pm) will increase chances to maximum to make the post on the top. To maximize the amount of points a post receives, we'd recommend the post to be categorized as show post and created around 12:00 (12:00 pm).